File: 2492-minimum-score-of-a-path-between-two-cities/solution.py
from collections import deque, defaultdict

class Solution:
    def minScore(self, n: int, roads: list[list[int]]) -> int:
        # Build adjacency list: node -> list of (neighbor, distance)
        graph = defaultdict(list)
        for u, v, d in roads:
            graph[u].append((v, d))
            graph[v].append((u, d))
            
        min_score = float('inf')
        visited = [False] * (n + 1)
        queue = deque([1])
        visited[1] = True
        
        # BFS to find all reachable nodes and the min edge weight in the component
        while queue:
            curr = queue.popleft()
            
            for neighbor, dist in graph[curr]:
                # Update global minimum score encountered
                min_score = min(min_score, dist)
                
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
                    
        return min_score

File: 2492-minimum-score-of-a-path-between-two-cities/solution.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

class Solution {
public:
    int minScore(int n, std::vector<std::vector<int>>& roads) {
        // Build adjacency list: node -> vector of {neighbor, distance}
        std::vector<std::vector<std::pair<int, int>>> adj(n + 1);
        for (const auto& road : roads) {
            adj[road[0]].push_back({road[1], road[2]});
            adj[road[1]].push_back({road[0], road[2]});
        }

        int min_score = INT_MAX;
        std::vector<bool> visited(n + 1, false);
        std::queue<int> q;

        // Start traversal from city 1
        q.push(1);
        visited[1] = true;

        while (!q.empty()) {
            int curr = q.front();
            q.pop();

            for (auto& edge : adj[curr]) {
                int neighbor = edge.first;
                int dist = edge.second;
                
                // Update the minimum score found so far
                min_score = std::min(min_score, dist);

                // If neighbor hasn't been visited, add to queue
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }

        return min_score;
    }
};